File: arrays/medium/53MaximumSubArray/solution.py
# solution.py
def max_subarray(nums):
    cur = max_so_far = nums[0]
    for x in nums[1:]:
        cur = x if cur < 0 else cur + x
        max_so_far = max(max_so_far, cur)
    return max_so_far


if __name__ == "__main__":
    print(max_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))

File: arrays/medium/53MaximumSubArray/solution.go
// solution.go
package main

import "fmt"

func maxSubArray(nums []int) int {
	// Kadane's algorithm
	maxSoFar := nums[0]
	cur := nums[0]
	for i := 1; i < len(nums); i++ {
		if cur < 0 {
			cur = nums[i]
		} else {
			cur += nums[i]
		}
		if cur > maxSoFar {
			maxSoFar = cur
		}
	}
	return maxSoFar
}

func main() {
	fmt.Println(maxSubArray([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}))
}
